File: resources/Rox/scripts/GridRole.js

GridRole = Class({
    Extends: RoleScript,

    roleName: "GridRole",
        
    init: function() {
        // TODO Initialise your object. Note you can't access this.actor yet.
        this.speed = directorScript.squareSize; // Number of pixels to move per tick when dx or dy are non zero.
        this.dx = 0; // -1, 0 or 1 depending on direction of travel
        this.dy = 0; // -1, 0 or 1 depending on direction of travel.
        this.between = 0; // Distance travelled between squares (0..squareSize)
        this.squashed = false; // Set to false when started to move and set to true just before onInvaded is called.
        // this.square = null; // The square I currently occupy.
    },

    tick: function() {

        if ( (this.dx != 0) || (this.dy != 0) ) {

            this.between += this.speed;

            if ( ! this.squashed ) {
                if (this.between >= this.square.grid.squareSize / 2) {
                    this.squashed = true;
                    var squashing = this.findLocalRole( this.dx, this.dy );
                    squashing.onInvaded( this );
                }
            }
                        
            this.actor.moveBy( this.dx * this.speed, this.dy * this.speed );

            if (this.between >= this.square.grid.squareSize) {
                // Correct for any overshoot
                var overshoot = this.between - this.square.grid.squareSize;
                this.actor.moveBy( this.dx * - overshoot, this.dy * - overshoot );
                
                // Stop the movement and occupy the new square.
                var dx = this.dx; // Remember the direction I was travelling as its needed after dx,dy are reset.
                var dy = this.dy;
                this.between = 0;
                this.dx = 0;
                this.dy = 0;
                this.occupySquare( this.findLocalSquare( dx, dy ) );
                this.onArrived(dx, dy);
            }
        }  
    },
    
    // Tidy up when dead. Make sure that the square I'm occupying and the square I'm entring no longer
    // reference me.
    onDeath: function() {
        if ( this.isMoving() ) {
            var sqaure = this.findLocalSquare( this.dx, this.dy );
            if ( square && square.entrant == this ) {
                square.entrant = null;
            }
        }
        if ( this.square ) {
            this.square.occupant = EmptyGridRole.instance;
        }
        this.actor.kill();
    },
    
    occupySquare: function( newSquare ) {
        // It is possible for another GridRole to have occupied my old square before I have completely vacated it,
        // in which case, do not interfer. Otherwise, reset the occupant to the global "Empty" object.
        if (this.square.occupant == this) {
            this.square.occupant = EmptyGridRole.instance;
        }
        
        if (newSquare == null) {
            this.actor.kill();
        } else {
            newSquare.occupant = this;
            newSquare.entrant = null;
        }
        
        this.square = newSquare;
    },
    
    
    // Boiler plate code - no need to alter it.
    getProperties: function() {
        return GridRole.properties;
    },
    
    // Called by Level.js just after a level has been loaded, and must also be called manually if you create
    // new actors on the grid dynamically.
    // Uses the actors position to calculate the correct square to occupy.
    placeOnGrid: function( grid ) {
        this.square = grid.getSquareByPixel( this.actor.getX(), this.actor.getY() );
        if (this.square != null) {
            this.square.occupant = this;
        } else {
            stdout.println( "Failed to find square! " + this.actor.getX() + "," + this.actor.getY() );
            this.actor.getAppearance().setAlpha( 128 );
        }
    },
    
    // Finds the square
    findLocalSquare: function( dx, dy ) {
        var x = this.square.x + dx;
        var y = this.square.y + dy;
        return this.square.grid.getSquare(x, y);
    },
    
    // Finds the square
    findLocalRole: function( dx, dy ) {
        var x = this.square.x + dx;
        var y = this.square.y + dy;
        return this.square.grid.getGridRole(x, y);
    },
    
    // Finds the GridRole in the given direction. If this lands us outside the whole grid, then a global "edge"
    // is returned, which is solid and unchanging. If there is nothing at that position, then a global "empty"
    // is returned.
    // When objects are moving between squares, the result is a little more complex.
    // If there is an entrant, then return that, not the occupier (which will move out, or be squashed).
    // However, if the occupant is leaving, and will be gone by the time I could enter, then don't return the occupant, 
    // and instead the global EmptyGridRole.instance.
    look: function( dx, dy, speed ) {
        this.debug( "Look " + dx + "," + dy );
        if (!speed) {
            speed = this.speed;
        }
        
        var square = this.findLocalSquare(dx, dy);
        // If there is an object entering, then this is the one we care about.
        var entrant = (square == null) ? null : square.entrant;
        this.debug( "Square : " + square + " entrant : " + entrant );
        if (entrant != null) {
            this.debug( "Using entrant because there is one" );
            return entrant;
        }
        
        var occupier = this.findLocalRole( dx, dy );
        if ( ! occupier.isMoving() ) {
            this.debug( "Using occupier because not moving" );
            return occupier;
        }
        
        // Calculate how long (in frames) it will take for the occupant to leave the square, and
        // for me to enter the square.
        var willLeaveTicks = (this.square.grid.squareSize - occupier.between) / occupier.speed;
        var willEnterTicks = this.square.grid.squareSize / speed;

        // this.debug( "Will leave : " + willLeaveTicks + " vs will enter : " + willEnterTicks );

        // If it will leave before I get half way, then ignore it.
        if (willLeaveTicks < willEnterTicks ) {
            this.debug( "Using EMPTY because the occupier will move out" );
            return EmptyGridRole.instance;
        }
        this.debug( "Using occupier because it isn't moving fast enough to ignore" );
        return occupier;
    },
    
    debug: function( message ) {
        return;
        if ( (this.square.x == 1) && (this.square.y == 1) ) {
            stdout.println( message );
        }
    },
    
    lookEast: function(speed) {
        return this.look( 1, 0, speed );
    },
    
    lookSouth: function(speed) {
        return this.look( 0,-1, speed );
    },
    
    lookWest: function(speed) {
        return this.look( -1, 0, speed );
    },
    
    lookNorth: function(speed) {
        return this.look( 0, 1, speed );
    },
    
    lookSouthEast: function(speed) {
        return this.look( 1, -1, speed );
    },
    
    lookSouthWest: function(speed) {
        return this.look( -1,-1 );
    },
    
    lookNorthWest: function(speed) {
        return this.look( -1, 1, speed );
    },
    
    lookNorthEast: function(speed) {
        return this.look( 1, 1, speed );
    },
    
    

    move: function(dx, dy) {
        if (this.isMoving()) {
            out.println( "Already moving. Ignored");
        } else {
            this.squashed = false;
            this.dx = dx;
            this.dy = dy;
            this.between = 0;
            this.findLocalSquare( dx, dy ).entrant = this;
        }
    },
    
    moveEast: function() {
        this.move( 1, 0 );
    },
    
    moveSouth: function() {
        this.move( 0, -1);
    },
    
    moveWest: function() {
        this.move( -1, 0 );
    },
    
    moveNorth: function() {
        this.move( 0, 1 );
    },
    

    pushed: function(pusher, dx, dy, force) {
        // Default is to be immovable.
        return false;
    },

    isEmpty: function() {
        return false;
    },
    
    isMoving: function() {
        return (this.dx != 0) || (this.dy != 0);
    },
    
    
    // Called after the object has finished travelling from one square to another.
    // dx,dy : The direction the invader was travelling ) {
    onArrived: function(invader, dx, dy ) { 
        // Do nothing.
    },
    
    // Called when a GridRole is halfway through encrouching into my teritory.
    onInvaded: function( invader ) {
        // Do nothing
    }
    
});

import("Grid.js");
import("EmptyGridRole.js");

GridRole.properties = new java.util.ArrayList();
// TODO Define each editable property like so :
// GridRole.properties.add( new itchy.property.StringProperty("myString").label("My String") );
// GridRole.properties.add( new itchy.property.IntegerProperty("myInteger").label("My Integer") );
// GridRole.properties.add( new itchy.property.DoubleProperty("myDouble").label("My Double") );
// GridRole.properties.add( new itchy.property.BooleanProperty("myBoolean").label("My Boolean") );
// GridRole.properties.add( new itchy.property.RGBAProperty("myColor").label("My Colour") );
// Each property should be initialised to a default value in the "init" method.

File: resources/Pyrox/scripts/gridRole.py
from uk.co.nickthecoder.itchy import Role
from uk.co.nickthecoder.itchy import AbstractRole
from uk.co.nickthecoder.itchy.util import ClassName
from uk.co.nickthecoder.itchy.extras import Fragment
from uk.co.nickthecoder.itchy.role import ExplosionBuilder
from uk.co.nickthecoder.itchy.role import TalkBuilder

from java.util import ArrayList

from grid import Grid

properties = ArrayList()

class GridRole(AbstractRole) :

    def __init__(self) :
        self.square = None
        self.role = self # So that the look method can return a GridRole or a Comound, and both have a "role" attribute.
        self.alternateRole = None # Same reason as self.role. This should always be None
        
        self.speed = 4 # The default speed for this object.
        
        # Each GridRole object knows when its tick was last called. Used by GridStage to ensure it
        # doesn't tick a role more than once per frame.
        self.latestTick = -1
        
        self.idle = True
        self.talkActor = None

        self.talkX = 70
        self.talkY = 10        


    def onBirth( self ) :
        Fragment().actor(self.actor).pieces( 10 ).createPoses( "fragment" )
        self.addTag( "explodable" )
    
    def onAttach( self ) :
        pass
        
    def talk( self, message ) :
        if self.talkActor :
            self.talkActor.kill()

        if message.startswith( "_"  ) :
            message = self.actor.costume.getString( message[1:], message[1:] )
        else :
            message = message.replace("\\n","\n")

        talk = TalkBuilder(self.actor) \
            .style("talk-style") \
            .offset(self.talkX, self.talkY).alignment( 1, 1 ) \
            .text(message)
        self.adjustTalk( talk );
        
        self.talkActor = talk.create().getActor()
        
        self.talkActor.getStage().addTop(self.talkActor)
        self.talkActor.setCostume( self.getActor().getCostume() )
        self.talkActor.event("talk-fade")
               
    def adjustTalk( self, talkBuilder ) :
        pass;

    def moveTo( self, x, y ) :
   
        grid = self.square.grid if self.square else None
            
        self.removeFromGrid()
        self.getActor().moveTo(x, y)
        if grid :
            self.placeOnGrid(grid)

    # Called by Level.js just after a level has been loaded, and must also be called manually if you create
    # new actors on the grid dynamically.
    # Uses the actors position to calculate the correct square to occupy.
    def placeOnGrid( self, grid ) :
    
        self.square = grid.getSquareByPixel( self.getActor().getX(), self.getActor().getY() )
        if self.square != None :
            self.square.occupant = self
            self.onPlacedOnGrid()
        else :
            print "Failed to find square! ", self.actor.getX(), ",", self.actor.getY()
            self.getActor().getAppearance().setAlpha( 128 )

    def onPlacedOnGrid( self ) :
        pass
    
    # Finds the square
    def findLocalSquare( self, dx, dy ) :
        x = self.square.x + dx
        y = self.square.y + dy
        return self.square.grid.getSquare(x, y)
    
    # Finds the square
    def findLocalRole( self, dx, dy ) :
        x = self.square.x + dx
        y = self.square.y + dy
        
        return self.square.grid.getGridRole(x, y)
    
    # Finds the GridRole in the given direction. If this lands us outside the whole grid, then a global "edge"
    # is returned, which is solid and unchanging. If there is nothing at that position, then a global "empty"
    # is returned.
    # When objects are moving between squares, the result is a little more complex.
    # If there is an entrant, then return that, not the occupier (which will move out, or be squashed).
    # However, if the occupant is leaving, and will be gone by the time I could enter, then don't return the occupant, 
    # and instead the global EmptyGridRole.instance.
    # In addition, each square has an alternateOccupant, and this is returned, when otherwise the result would have
    # been Empty.
    # One final, and tricky complication, a square can have an occupant and an alternateOccupant. If both are found,
    # then instead of a GridRole and Compound is returned. Therefore you cannot assume the result from look is a
    # role, if you want the role, add ".role". e.g. : foo.look( 1,0 ).role.onExplode()
    def look( self, dx, dy, speed=None ) :

        if speed is None :
            speed = self.speed
        
        square = self.findLocalSquare(dx, dy)
        # If there is an object entering, then this is the one we care about.
        entrant = None if (square is None) else square.entrant
        if entrant :
            return entrant
        
        if square and square.alternateOccupant :
            return square.getOccupant()
             
            
        occupier = self.findLocalRole( dx, dy )
        if not occupier.isMoving() :
            return occupier

        # Calculate how long (in frames) it will take for the occupant to leave the square, and
        # for me to enter the square.
        willLeaveTicks = (self.square.grid.squareSize - occupier.between) / occupier.speed
        willEnterTicks = self.square.grid.squareSize / speed

        isOpposite = dx == -occupier.dx and dy == -occupier.dy
        
        # If it will leave before I get half way, then ignore it.
        if willLeaveTicks < willEnterTicks and not isOpposite :
            if square.alternateOccupant :
                return square.alternateOccupant
            else :
                return square.grid.getEmpty()

        return occupier
        
    def lookDirection(self, direction, speed=None) :

        return self.look( self.getDeltaX(direction), self.getDeltaY(direction), speed )


    def lookEast(self, speed=None) :
        return self.look( 1, 0, speed )
    
    def lookSouth(self, speed=None) :
        return self.look( 0,-1, speed )
    
    def lookWest(self,speed=None) :
        return self.look( -1, 0, speed )
    
    def lookNorth(self,speed=None) :
        return self.look( 0, 1, speed )
    
    def lookSouthEast(self, speed=None) :
        return self.look( 1, -1, speed )
    
    def lookSouthWest(self, speed=None) :
        return self.look( -1,-1, speed )
    
    def lookNorthWest(self, speed=None) :
        return self.look( -1, 1, speed )
    
    def lookNorthEast(self, speed=None) :
        return self.look( 1, 1, speed )
    
    
    def canShove( self, pusher, dx, dy, speed, force ) :
        # Default is to be immovable.
        return False

    def shove( self, pusher, dx, dy, speed ) :
        pass
        
    def isEmpty( self ) :
        return False
    
    def isMoving( self ) :
        return False

    def onExplode( self ) :
        self.explode()

    def explode( self ) :        
    
        ExplosionBuilder(self.actor) \
            .projectiles(10) \
            .gravity(-0.2).fade(0.9, 3.5).speed(0.1, 1.5).vy(5) \
            .pose("fragment") \
            .create()

        self.actor.deathEvent("explode")


    # Called when a GridRole is halfway through encrouching into my teritory.
    def onHalfInvaded( self, invader ) :
        pass
                
    # Called when a GridRole is just about to take over my square.
    # The only way to prevent me being booted out is to override this method, and remove the invader
    # from the grid instead.
    def onInvaded( self, invader ) :
        self.removeFromGrid()
        
    # Tidy up when dead. Make sure that the square I'm occupying and the square I'm entring no longer
    # reference me.
    def onDeath( self ) :
            
        self.removeFromGrid()
    
    def removeFromGrid( self ) :
        if self.isMoving() :
            square = self.findLocalSquare( self.dx, self.dy )
            if square and (square.entrant == self) :
                square.entrant = None
                
        if self.square :
            if self.square.occupant == self :
                self.square.occupant = None
            elif self.square.alternateOccupant == self :
                self.square.alternateOccupant = None

        self.square = None
        self.dx = 0
        self.dy = 0

    # Each square has an alternative occupant, so that two roles can share the same square.
    # This methods downgrades a role from being the normal occupant to the alternate occupant.
    # Must NOT be used for movable objects
    def makeAlternateOccupant( self ) :
    
        if self.square.occupant != self :
            return
        self.square.occupant = None

        if self.square.alternateOccupant :
            self.square.alternateOccupant.removeFromGrid()        
        
        self.square.alternateOccupant = self;
    
    def unmakeAlternateOccupant( self ) :
        if self.square.alternateOccupant != self :
            print "Not an alternateOccupant!"
            return
        if self.square.occupant != None :
            print "Not empty"
            return
        self.square.alternateOccupant = None
        self.square.occupant = self


    @staticmethod
    def getDirectionAbbreviation( direction ) :
        return [ "E","N","W","S" ][ direction ]

    @staticmethod
    def getCompassAbbreviation( dx, dy ) :
        return ["W", "", "E" ][dx + 1] + ["S", "", "N"][dy + 1]

    @staticmethod
    def getDeltaX( direction ) :
        return [1,0,-1,0][direction]

    @staticmethod
    def getDeltaY( direction ) :
        return [0,1,0,-1][direction]


    # Boiler plate code - no need to change this
    def getProperties(self):
        return properties

    # Boiler plate code - no need to change this
    def getClassName(self):
        return ClassName( Role, self.__module__ + ".py" )


    def toString(self) :
        return self.__str__()
        
    def __str__(self) :
        return self.__class__.__name__ + " @ " + str(self.square.x) + "," + str(self.square.y)

    

